hibernate-core/src/main/java/org/hibernate/internal/util/collections/InstanceIdentityMap.java

```
@@ -0,0 +1,377 @@
+		if ( entry != null && entry.getKey() == key ) {
+			return entry.getValue();
+		}
+		return null;
+	}
+
+	/**
+	 * @inheritDoc
+	 * @implNote This only works for {@link InstanceIdentity} keys, and it's inefficient
+	 * since we need to do a type check. Prefer using {@link #get(int, Object)}.
+	 */
+	@Override
+	public @Nullable V get(Object key) {
+		if ( key instanceof InstanceIdentity instance ) {
+			return get( instance.$$_hibernate_getInstanceId(), instance );
+		}
+		throw new ClassCastException( "Provided key does not support instance identity" );
+	}
+
+	@Override
+	public @Nullable V put(K key, V value) {
+		if ( key == null ) {
+			throw new NullPointerException( "This map does not support null keys" );
+		}
+
+		final int instanceId = key.$$_hibernate_getInstanceId();
+		final Map.Entry<K, V> old = set( instanceId, new AbstractMap.SimpleImmutableEntry<>( key, value ) );
+		if ( old == null ) {
+			size++;
+			return null;
+		}
+		else {
+			return old.getValue();
+		}
+	}
+
+	/**
+	 * Removes the mapping for an instance id from this map if it is present (optional operation).
+	 *
+	 * @param instanceId the instance id whose associated value is to be returned
+	 * @param key key instance to double-check instance equality
+	 * @return the previous value associated with {@code instanceId}, or {@code null} if there was no mapping for it.
+	 * @implNote This method accesses the backing array with the provided instance id, but performs an instance
+	 * equality check ({@code ==}) with the provided key to ensure it corresponds to the mapped one
+	 */
+	public @Nullable V remove(int instanceId, Object key) {
+		final Page<Map.Entry<K, V>> page = getPage( instanceId );
+		if ( page != null ) {
+			final int pageOffset = toPageOffset( instanceId );
+			final Map.Entry<K, V> entry = page.set( pageOffset, null );
+			// Check that the provided instance really matches with the key contained in the map
+			if ( entry != null ) {
+				if ( entry.getKey() == key ) {
```

Maybe throw an exception when the keys don't match?